src/quorum/majority.rs

```diff
@@ -130,6 +131,82 @@ impl Configuration {
             VoteResult::Lost
         }
     }
+
+    /// Describe returns a (multi-line) representation of the commit indexes for the
+    /// given lookuper.
+    /// Including `Index`,`Id` and the number of smaller index (represented as the bar)
+    ///
+    /// Print `?` if `Index` is not exist.
+    ///
+    /// e.g.
+    /// ```txt
+    ///             idx
+    /// x>          100 (id=1)
+    /// xx>         101 (id=2)
+    /// >            99 (id=3)
+    /// 100
+    /// ```
+    pub fn describe(&self, l: &impl AckedIndexer) -> String {
+        let n = self.voters.len();
+        if n == 0 {
+            return "<empty majority quorum>".to_string();
+        }
+
+        struct Tup {
+            id: u64,
+            idx: Option<Index>,
+            // length of bar displayed for this Tup
+            bar: usize,
+        }
+
+        // Below, populate .bar so that the i-th largest commit index has bar i (we
+        // plot this as sort of a progress bar). The actual code is a bit more
+        // complicated and also makes sure that equal index => equal bar.
+
+        let mut info = vec![];
+
+        for id in self.raw_slice() {
+            let idx = l.acked_index(id);
+            info.push(Tup { id, idx, bar: 0 })
+        }
+
+        info.sort_by(|a, b| match a.idx.cmp(&b.idx) {
+            Ordering::Equal => a.id.cmp(&b.id),
+            ordering => ordering,
+        });
+
+        for i in 0..n {
+            if i > 0 && info[i - 1].idx < info[i].idx {
+                info[i].bar = i;
+            }
+        }
+
+        info.sort_by(|a, b| a.id.cmp(&b.id));
+
+        let mut buf = String::new();
+        buf.push_str(&(" ".repeat(n) + format!("  {:>5}\n", "idx").as_str()));
+
+        for tup in info {
+            let string;
+            if let Some(idx) = tup.idx {
+                string = "x".repeat(tup.bar)
+                    + ">"
+                    + " ".repeat(n - tup.bar).as_str()
+                    + format!(" {:>5}    (id={})\n", format!("{}", idx), tup.id).as_str();
```

Comment:
Redundant code, can be extracted as `write!(buf, " {:>5}    (id={})\n", tup.idx.unwrap_or_default(), tup.id).unwrap()`.

Reply:
I still use branch to determine the prefix of the string

Reply:
I fail to get width with 
`writeln!(buf, " {:>5} (id={})", idx, tup.id)
                        .expect("Error occurred while trying to write in String");
                ` and 
`writeln!(buf, " {:>5} (id={})", Index::default(), tup.id)
                        .expect("Error occurred while trying to write in String");
                `

Reply:
It works if passing `format("{}", idx)`

Reply:
Can you talk more about the error you meet? Or you can show me a picture in slack.

Reply:
the following is the result with this code `writeln!(buf, " {:>5}    (id={})", idx, tup.id)`, 
the green one is the expected output.
![](https://i.imgur.com/ww3IpOp.png)

Reply:
with this code `writeln!(buf, " {:>5}    (id={})", format!("{}", idx), tup.id)`, I can get the green output

Reply:
It seems the problem may be in `Display`
https://stackoverflow.com/questions/56826464/why-is-the-width-ignored-for-my-custom-formatter-implementation